src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,511 @@
+                    RENDERER_LOG_ERROR(
+                        "this message should not be logged"
+                        "using default value \"equisolid_angle\".");
+                    theta2 = 2 * asin(tan_theta1 / 2.0);
+            }
+
+            const Transformd rot2 = Transformd(Matrix4d::make_rotation_y(theta2 - theta1));
+
+            dir = rot2.vector_to_parent(dir);
+            dir = rot.vector_to_local(dir);
+        }
+
+        void spawn_ray(
+            SamplingContext&    sampling_context,
+            const Dual2d&       ndc,
+            ShadingRay&         ray) const override
+        {
+
+            // Initialize the ray.
+            initialize_ray(sampling_context, ray);
+
+            // Retrieve the camera transform.
+            Transformd scratch;
+            const Transformd& transform =
+                m_transform_sequence.evaluate(ray.m_time.m_absolute, scratch);
+
+            // Compute ray origin and direction.
+            ray.m_org = transform.get_local_to_parent().extract_translation();
+            ray.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(ndc.get_value())));
+
+            transform_ray_dir(ray.m_dir);
+
+            // Compute ray derivatives.
+            if (ndc.has_derivatives())
+            {
+                const Vector2d px(ndc.get_value() + ndc.get_dx());
+                const Vector2d py(ndc.get_value() + ndc.get_dy());
+                ray.m_rx.m_org = ray.m_org;
+                ray.m_ry.m_org = ray.m_org;
+                ray.m_rx.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(px)));
+                transform_ray_dir(ray.m_rx.m_dir);
+                ray.m_ry.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(py)));
+                transform_ray_dir(ray.m_ry.m_dir);
+                ray.m_has_differentials = true;
+            }
+        }
+
+        bool connect_vertex(
+            SamplingContext&    sampling_context,
+            const float         time,
+            const Vector3d&     point,
+            Vector2d&           ndc,
+            Vector3d&           outgoing,
+            float&              importance) const override
+        {
+            // Project the point onto the film plane.
+            if (!project_point(time, point, ndc))
+                return false;
+
+            // The connection is impossible if the projected point lies outside the film.
+            if (ndc[0] < 0.0 || ndc[0] >= 1.0 ||
+                ndc[1] < 0.0 || ndc[1] >= 1.0)
+                return false;
+
+            // Retrieve the camera transform.
+            Transformd scratch;
+            const Transformd& transform = m_transform_sequence.evaluate(time, scratch);
+
+            // Compute the outgoing direction vector in world space.
+            outgoing = point - transform.get_local_to_parent().extract_translation();
+
+            // Compute the emitted importance.
+            const Vector3d film_point = ndc_to_camera(ndc);
```

Comment:
You will need to properly compute importance for each type of projection.

Reply:
Where can I get some informations about emitted importance? Is it included in PBRT?

Reply:
There are definitely some good information about this topic in PBRT (look for importance sampling). The "importance" that you must compute is in fact the probability that the world space direction `outgoing` get chosen by your `spawn_ray()` method.

Reply:
Note that `connect_vertex()` is not called by the unidirectional path tracer. To see the problem and test your implementation of `connect_vertex()`, you will need to switch to SPPM. I also suggest using a much simpler scene (such as the Cornell Box) for your tests.

If you render the Cornell Box with SPPM, with your new camera and with the current implementation of `connect_vertex()`, you should notice some brightness issues: maybe some areas will be slightly darker/brighter than they should. It can be subtle, which adds to the difficulty of writing a correct implementation of `connect_vertex()`.

Reply:
> The "importance" that you must compute is in fact the probability 

Since `transform_ray_dir()` is kind of one-to-one mapping function, isn't it works well?